### ISSUE TITLE
Multiple-systems configuration always inherit default environment - Finchley.M8/Spring Boot 2.0

### ISSUE THREAD

**Issue by wildloop** [spring-cloud/spring-cloud-stream#1273]
[My configuration](https://github.com/wildloop/streams-demo/blob/sb2.0/src/main/resources/application-2vh.yml) based on example from documentation ["Connecting to Multiple Systems"](https://docs.spring.io/spring-cloud-stream/docs/current/reference/htmlsingle/#multiple-systems):

```json
spring:
  rabbitmq:
    username: demo
    password: demo
    virtualHost: demo
  cloud:
    stream:
      binders:
        demoin:
          type: rabbit
          defaultCandidate: false
          inheritEnvironment: false
          environment:
            spring:
              rabbitmq:
                username: demo
                password: demo
                virtualHost: demo-a
        demoout:
          type: rabbit
          defaultCandidate: false
          inheritEnvironment: false
          environment:
            spring:
              rabbitmq:
                username: demo
                password: demo
                virtualHost: demo-b
      bindings:
        input:
          binder: demoin
          destination: in-demo-a-dest
          group: in-demo-a-group
        output:
          binder: demoout
          destination: out-demo-b-dest
          group: out-demo-b-group
```

This config should cause two connections as binders:
- demoin: virtual-host demo-a
- demoout: virtual-host demo-b

but we get (log):
- o.s.a.r.c.CachingConnectionFactory: Created new connection: rabbitConnectionFactory#2121d1f9:0/SimpleConnection@4a481728 [delegate=amqp://demo@127.0.0.1:5672/**demo**, localPort= 51692]
- o.s.a.r.c.CachingConnectionFactory: Created new connection: rabbitConnectionFactory#76bf1bb8:0/SimpleConnection@1b482cbf [delegate=amqp://demo@127.0.0.1:5672/**demo**, localPort= 51694]

and RabbitMQ Admin UI screen:

![rabbit-vh](https://user-images.githubusercontent.com/2461648/36950871-bcb3e468-1ffc-11e8-9b56-f4274f73cc6c.png)

[This configuration works](https://github.com/wildloop/streams-demo/blob/master/src/main/resources/application-2vh.yml) in the Spring Cloud Edgware.SR2 and Spring Boot 1.5.9.RELEASE

**Comment by olegz** [comment]
@wildloop thank you for pointing this out. It appears to be related to the following issues https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/issues/326
https://github.com/spring-cloud/spring-cloud-stream/issues/1124

Not sure yet what could be the problem, can't recall any work related to this, but it's great we are catching it now.

**Comment by garyrussell** [comment]
Related to https://stackoverflow.com/questions/49089915/how-to-set-custom-name-for-rabbitmq-connection/49090085#49090085

**Comment by olegz** [comment]
@wildloop I was able to reproduce it by switching to boot 1.5.9 and Edgware.SR2 , so in a way it's a good news.

**Comment by wildloop** [comment]
@olegz How did you do that?
I repeated the test and it worked ok for me with SB1.5.9/Edgware.SR2...
In [this repo's master branch](https://github.com/wildloop/streams-demo/blob/master/pom.xml) I have `1.5.9.RELEASE` and `Edgware.SR2`. When I run app with profile [`2vh`](https://github.com/wildloop/streams-demo/blob/master/src/main/resources/application-2vh.yml) then I get connections to virtual hosts that I want to:  **demo-a** and **demo-b**:

- Created new connection: rabbitConnectionFactory#e95595b:0/SimpleConnection@36aab105 [delegate=amqp://demo@127.0.0.1:5672/**demo-b**, localPort= 61920]
- Created new connection: rabbitConnectionFactory#f2b90fc:0/SimpleConnection@24c685e7 [delegate=amqp://demo@127.0.0.1:5672/**demo-a**, localPort= 61921]

**Comment by olegz** [comment]
Sorry, I misspoke. I meant that it works as expected with 1.5.9/Edgware but not with 2.0.0/Finchley. Basically conformed your case. Sorry for the confusion. 
Btw, we're tracking it as boot issue, but in any event will have some fix on our end before GA, so stay tuned

**Comment by wildloop** [comment]
ok, I already thought we found [The Matrix bug](http://www.businessinsider.com/bank-of-america-wonders-about-the-matrix-2016-9) ;)

**Comment by olegz** [comment]
Quick update:
So based on the earlier issue (https://github.com/spring-cloud/spring-cloud-stream/issues/1136) from @dsyer  we needed to change our `BinderProperties` as such:
```
@Deprecated
public void setEnvironment(Properties environment){..}
```
to 
```
public void setEnvironment(Map<Object, Object> environment) {..}
```

Now, the boot's `JavaBeanBinder` inconsistently invokes either of those operations. However the issue is that when it invokes the first one it passes environment as
`spring.rabbitmq.password=demouser, spring.rabbitmq.virtualHost=demo-a, spring.rabbitmq.username=demouser`
the later:
`spring={rabbitmq={username=demouser, password=demouser, virtualHost=demo-a}}`
Basically nested maps aren't flattened to `.` delimited properties, which affects our subsequent parsing and falsely inherits the environment

Knowing it, we will have a fix/workaround on our end for GA, but given the inconsistent behavior in boot we have communicated it to boot team

**Comment by olegz** [comment]
@wildloop so it is fixed with b7f7e3b. Give it a shot and let us know if you still see the issue

**Comment by olegz** [comment]
@wildloop if you can could you please confirm that this has been fixed? You'd need to pull the snapshot.
Thank you

**Comment by wildloop** [comment]
Now is ok - `SB 2.0.0` / `Spring Cloud Streams 2.0.0.BUILD-SNAPSHOT`:

- Created new connection: rabbitConnectionFactory#6c5ddccd:0/SimpleConnection@411a5965 [delegate=amqp://demo@127.0.0.1:5672/**demo-b**
- Created new connection: rabbitConnectionFactory#6a1ef65c:0/SimpleConnection@80b122b [delegate=amqp://demo@127.0.0.1:5672/**demo-a**

Thank you :)
